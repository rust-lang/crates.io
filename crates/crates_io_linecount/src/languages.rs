use tokei::LanguageType;

/// Determine if a language should be counted or ignored
pub fn should_ignore_language(lang: LanguageType) -> bool {
    matches!(
        lang,
        // Configuration and data files
        LanguageType::Json |
        LanguageType::Yaml |
        LanguageType::Toml |
        LanguageType::Xml |
        LanguageType::Ini |

        // Documentation
        LanguageType::Markdown |
        LanguageType::Text |
        LanguageType::ReStructuredText |
        LanguageType::AsciiDoc |
        LanguageType::Org |

        // Build system files
        LanguageType::Makefile |
        LanguageType::CMake |
        LanguageType::Dockerfile |
        LanguageType::Autoconf |
        LanguageType::MsBuild |
        LanguageType::Meson |
        LanguageType::Scons |
        LanguageType::Bazel |
        LanguageType::Nix |

        // Shell scripts (debatable, but often just build/deploy automation)
        LanguageType::Batch |
        LanguageType::PowerShell |

        // Other non-programming files
        LanguageType::Svg |
        LanguageType::Hex |
        LanguageType::Protobuf |
        LanguageType::Thrift
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_should_ignore_language() {
        // Should count programming languages
        assert!(!should_ignore_language(LanguageType::Rust));
        assert!(!should_ignore_language(LanguageType::JavaScript));
        assert!(!should_ignore_language(LanguageType::Html));
        assert!(!should_ignore_language(LanguageType::Css));

        // Should skip config/data files
        assert!(should_ignore_language(LanguageType::Json));
        assert!(should_ignore_language(LanguageType::Yaml));
        assert!(should_ignore_language(LanguageType::Toml));
        assert!(should_ignore_language(LanguageType::Markdown));
    }
}
