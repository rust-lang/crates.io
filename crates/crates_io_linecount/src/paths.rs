use crate::languages::should_ignore_language;
use std::path::Path;
use tokei::LanguageType;

#[derive(Debug, Clone, Copy)]
pub struct PathDetails {
    is_benchmark: bool,
    is_example: bool,
    is_hidden: bool,
    is_test: bool,
    language_type: Option<LanguageType>,
}

impl PathDetails {
    pub fn from_path(path: &Path) -> Self {
        let path_str = path.to_string_lossy().to_lowercase();

        let is_benchmark = path_str.contains("benches/") || path_str.contains("benchmark/");
        let is_example = path_str.contains("examples/");
        let is_test = path_str.contains("tests/")
            || path_str.contains("test/")
            || path_str.contains("testing/");

        let is_hidden = path
            .file_name()
            .map(|filename| filename.to_string_lossy().starts_with('.'))
            .unwrap_or(false);

        let language_type = path
            .extension()
            .and_then(|ext| ext.to_str())
            .and_then(LanguageType::from_file_extension);

        Self {
            is_benchmark,
            is_example,
            is_hidden,
            is_test,
            language_type,
        }
    }

    /// Determine if the file should be ignored for line counting purposes
    /// because it is a benchmark, example, hidden, or test file.
    pub fn should_ignore(&self) -> bool {
        self.is_benchmark || self.is_example || self.is_hidden || self.is_test
    }

    /// Get the actual detected language type, even if it should be ignored.
    pub fn actual_language_type(&self) -> Option<LanguageType> {
        self.language_type
    }

    /// Get the detected language type, returning `None` if no language was
    /// detected or if the language should be ignored (e.g., data files).
    pub fn language_type(&self) -> Option<LanguageType> {
        self.language_type.filter(|lt| !should_ignore_language(*lt))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_should_count_path() {
        assert_debug_snapshot!(PathDetails::from_path(Path::new("src/tests/mod.rs")));
        assert_debug_snapshot!(PathDetails::from_path(Path::new("tests/integration.rs")));
        assert_debug_snapshot!(PathDetails::from_path(Path::new("examples/basic.rs")));
        assert_debug_snapshot!(PathDetails::from_path(Path::new("benches/bench.rs")));
        assert_debug_snapshot!(PathDetails::from_path(Path::new("src/lib.rs")));
    }
}
