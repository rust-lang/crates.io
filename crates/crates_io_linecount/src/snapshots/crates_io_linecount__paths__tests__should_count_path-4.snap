---
source: crates/crates_io_linecount/src/paths.rs
expression: "PathDetails::from_path(Path::new(\"benches/bench.rs\"))"
---
PathDetails {
    is_benchmark: true,
    is_example: false,
    is_hidden: false,
    is_test: false,
    language_type: Some(
        Rust,
    ),
}
